bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java

```diff
@@ -966,8 +970,25 @@ public void run() {
                             forceWriteBatchBytesStats.registerSuccessfulValue(batchSize);
 
                             boolean shouldRolloverJournal = (lastFlushPosition > maxJournalSize);
-                            forceWriteRequests.put(createForceWriteRequest(logFile, logId, lastFlushPosition, toFlush, shouldRolloverJournal, false));
-                            toFlush = RecyclableArrayList.newInstance();
+                            if (syncData) {
+                                // Trigger data sync to disk in the "Force-Write" thread. Callback will be triggered after data is committed to disk
+                                forceWriteRequests.put(createForceWriteRequest(logFile, logId, lastFlushPosition, toFlush, shouldRolloverJournal, false));
+                                toFlush = RecyclableArrayList.newInstance();
+                            } else {
+                                // Data is already written on the file (though it might still be in the OS page-cache)
+                                lastLogMark.setCurLogMark(logId, lastFlushPosition);
+                                for (int i = 0; i < toFlush.size(); i++) {
+                                    cbThreadPool.execute((QueueEntry) toFlush.get(i));
+                                }
+
+                                toFlush.clear();
+                                if (shouldRolloverJournal) {
+                                    forceWriteRequests.put(createForceWriteRequest(logFile, logId, lastFlushPosition,
+                                            RecyclableArrayList.newInstance(), shouldRolloverJournal, false));
```

Comment:
I am not exactly sure if we also need do `toFlush = RecyclableArrayList.newInstance();` after it. So, does it make sense to have below condition?
```
if(syncData || shouldRolloverJournal) {
 // Trigger data sync to disk
  :
  forceWriteRequests.put(..)
} else {
  :
}
```